File: utils/generate-dot-file.lua
#!/usr/bin/env lua
-- SPDX-FileCopyrightText: 2020 Jonathan Poelen <jonathan.poelen@gmail.com>
-- SPDX-License-Identifier: MIT

-- load modules or show installation commands
_, argparse = xpcall(require, function(err) trace1 = err end, 'argparse')
_, xmlparser = xpcall(require, function(err) trace2 = err end, 'xmllpegparser')
if not argparse or not xmlparser then
    local versparam = '--lua-version=' .. _VERSION:match('Lua (.*)')
    io.stderr:write('Please install '
                    .. (argparse and '' or 'argparse')
                    .. (not argparse and not xmlparser and ' and ' or '')
                    .. (xmlparser and '' or 'xmllpegparser')
                    .. ':\n\n    luarocks --local ' .. versparam .. ' install'
                    .. (argparse and '' or ' argparse')
                    .. (xmllpegparser and '' or ' xmllpegparser')
                    .. '\n\n    eval "$(luarocks ' .. versparam .. ' path)"\n\n'
                    .. (trace1 or trace2) .. '\n')
    os.exit(1)
end

parser = argparse(arg[0],
                  "Dot file generator for xml syntax\n\nExample:\n    "
                  .. arg[0] .. " data/syntax/lua.xml | dot -T png -o image.png")

function parser:pattern(name, description)
    parser
    :option(name, description .. "\nCan be present several times.")
    :argname'<pattern>'
    :count'*'
end

parser:flag('-c --context-only', "Generates contexts without rules")
parser:flag('-r --replace-entities', "Evaluate html entities")
parser:pattern('-i --include', "Include only contexts that respect a (lua) pattern.")
parser:pattern('-e --exclude', "Exclude contexts that respect a (lua) pattern.")
parser:argument('syntax.xml', "Syntax Definition Files")

do
    -- same message for -h, --help and usage
    local help = parser:get_help()
    parser:usage(help)
    parser:help(help)
end

args = parser:parse()

excludes = args.exclude
-- '' is a pattern always found
includes = #args.include == 0 and {''} or args.include
contextOnly = args.context_only
replaceEntities = args.replace_entities and not contextOnly


do
    -- load file
    local f, err = io.open(args['syntax.xml'])
    if not f then
        io.stderr:write(err .. '\n')
        os.exit(2)
    end

    local content = f:read'*a'

    -- remove BOM
    if content:sub(1,3) == '\xef\xbb\xbf' then
        content = content:sub(4)
    end

    -- load document
    document, err = xmlparser.parse(content)
    if not document then
        io.stderr:write(err .. '\n')
        os.exit(3)
    end
end


local contexts
-- search language/highlighting/contexts
for k,nodes in pairs(document.children[1].children[1].children) do
    if (nodes.tag == 'contexts') then
        contexts = nodes.children
        break
    end
end

if not contexts then
    io.stderr:write('<contexts> not found\n')
    os.exit(4)
end

if replaceEntities then
    local entities = xmlparser.createEntityTable(document.entities)

    for _,ctx in pairs(contexts) do
        for _,rule in pairs(ctx.children) do
            for _,k in pairs({'String', 'char', 'char1'}) do
                if rule.attrs[k] then
                    rule.attrs[k] = xmlparser.replaceEntities(rule.attrs[k], entities)
                end
            end
        end
    end
end


local colorMap = {
    '"/set312/1"',
    '"lightgoldenrod1"',
    '"/set312/3"',
    '"/set312/4"',
    '"/set312/5"',
    '"/set312/6"',
    '"/set312/7"',
    '"/rdpu3/2"',
    '"/rdgy4/3"',
    '"/purd6/3"',
    '"/ylgn4/2"',
    '"/set26/6"',
}

-- returns a color which depends on the first 2 characters
function computeColor(name)
    local hash = name:byte(1)
    if #name ~= 1 then
        hash = hash + name:byte(2) * 25
    end
    return colorMap[hash % #colorMap + 1];
end

function matchPatterns(name, patterns)
    for _,v in pairs(patterns) do
        if name:find(v) then
            return true
        end
    end
    return false
end

function matchContext(name)
    return not matchPatterns(name, excludes) and matchPatterns(name, includes)
end

fmt = string.format

do
    local lpeg = require'lpeg'
    local P = lpeg.P
    local C = lpeg.C
    local S = lpeg.S
    local Cf = lpeg.Cf
    local Cc = lpeg.Cc
    local Cs = lpeg.Cs
    local Ct = lpeg.Ct

    -- #pop counter
    _countpop = Cf((P'#pop' * Cc(1))^1, function(a,b) return a+b end)
    -- cut by slice of 40 characters
    _wordwap = Ct(Cc('') * C(P(40) + P(1)^1)^0)
    -- replace " by \" and \ by \\
    _escape = Cs((S'"' / '\\"' + S'\\' / '\\\\' + 1)^0)
    -- extract context of #pop#pop#pop!Context
    _jumpctx = (1 - S'!')^1 * '!' * C(P(1)^1)
end

function labelize(name)
    local n = _countpop:match(name)
    if n and n > 1 then
        return fmt('#pop(%d)%s', n, name:sub(n * 4 + 1))
    end
    return name
end

-- convert {[k]={data, index}, ...} to {{k, data, index}, ...} sorted by index
-- for stable output
function kIndexedTableToTable(kt)
    local t = {}
    for k,x in pairs(kt) do
        t[#t+1] = {k, x[1], x[2]}
    end
    table.sort(t, function(a,b) return a[3] < b[3] end)
    return t
end

local sharp = string.byte('#',1)


if contextOnly then
    print('digraph G {')
    print('  compound=true;ratio=auto')

    for _,ctx in pairs(contexts) do
        local ctxName = ctx.attrs.name
        if matchContext(ctxName) then
            -- avoid multi arrow for ctx1 -> ctx2
            local kRuleContexts = {}
            for i,rule in ipairs(ctx.children) do
                kRuleContexts[rule.attrs.context or '#stay'] = {i,i}
            end

            if kRuleContexts['#stay'] and kRuleContexts[ctxName] then
                kRuleContexts[ctxName] = nil
            end

            local color = computeColor(ctxName)
            local escapedCtxName = _escape:match(ctxName)
            print(fmt('  "%s" [style=filled,color=%s]', escapedCtxName, color))

            for _,t in pairs(kIndexedTableToTable(kRuleContexts)) do
                local ruleContext, i = t[1], t[2]
                if ruleContext == '#stay' then
                    print(fmt('  "%s" -> "%s" [color=%s]',
                              escapedCtxName, escapedCtxName, color))
                elseif ruleContext:byte(1) == sharp then
                    local escapedRuleContext = _escape:match(ruleContext)
                    local nextContext = _jumpctx:match(escapedRuleContext)
                    print(fmt('  "%s" -> "%s!!%d" [color=%s];\n  "%s!!%d" [label="%s"];',
                              escapedCtxName,
                              escapedCtxName, i, color,
                              escapedCtxName, i, labelize(escapedRuleContext)))
                    if nextContext then
                        print(fmt('  "%s!!%d" -> "%s"',
                                  escapedCtxName, i, nextContext))
                    end
                else
                    print(fmt('  "%s" -> "%s" [color=%s]',
                          escapedCtxName, labelize(_escape:match(ruleContext)), color))
                end
            end
        end
    end

    print('}')

    return 0
end

function stringifyAttrs(t, attrs)
    local attr
    local s = ''
    for k,v in pairs(t) do
        if attrs[v] then
            attr = attrs[v]
            if #attr > 40 then
                attr = table.concat(_wordwap:match(attr),'\n')
            end
            s = s .. '  ' .. v .. ':' .. attr
        end
    end
    return s
end

function xmlBool(s)
    return s == 'true' or s == '1'
end

function printContextAttr(escapedOrigin, escapedCtxName, escapedNameAttr, style, color)
    if not escapedNameAttr then
        print(fmt('    "%s" -> "%s" [style=%s,color=%s];',
                  escapedOrigin, escapedCtxName, style, color))
    elseif escapedNameAttr:byte(1) == sharp then
        print(fmt('    "%s" -> "%s!!%s" [style=%s,color=%s];\n    "%s!!%s" [label="%s",color=%s];',
                  escapedOrigin,
                  escapedCtxName, escapedNameAttr, style, color,
                  escapedCtxName, escapedNameAttr, labelize(escapedNameAttr), color))
    end
end

function printLastTransition(escapedName, escapedCtxName, escapedNameAttr, color)
    if escapedNameAttr:byte(1) == sharp then
        local escapedLastCtx = _jumpctx:match(escapedNameAttr)
        if escapedLastCtx then
            print(fmt('  "%s!!%s" -> "%s" [style=dashed,color=%s];',
                      escapedCtxName, escapedNameAttr, escapedLastCtx, color))
        end
    else
        print(fmt('  "%s" -> "%s" [style=dashed,color=%s];',
                  escapedName, escapedNameAttr, color))
    end
end

local firstLineAttrs = {'attribute','String','char','char1'}
local secondLineAttrs = {'beginRegion','endRegion','lookAhead','firstNonSpace', 'column'}

print('digraph G {')
print('  compound=true;ratio=auto')
for ictx,ctx in pairs(contexts) do
    local ctxName = ctx.attrs.name
    if matchContext(ctxName) then
        local color = computeColor(ctxName)
        local escapedCtxName = _escape:match(ctxName)
        print(fmt('  subgraph cluster%d {', ictx))
        print(fmt('    "%s" [shape=box,style=filled,color=%s];', escapedCtxName, color))

        local name = ctxName
        local escapedName = escapedCtxName
        local kDot = {}
        for irule,rule in pairs(ctx.children) do
            local nextName = ctxName .. '!!' .. irule .. '!!' .. rule.tag
            local escapedNextName = _escape:match(nextName)
            local ruleContext = rule.attrs.context
            print(fmt('    "%s" -> "%s" [style=dashed,color=%s];',
                      escapedName, escapedNextName, color))
            name = nextName
            escapedName = escapedNextName

            local a = ''
            if rule.tag == 'IncludeRules' then
                a = '  ' .. ruleContext
            else
                if not rule.attrs.attribute then
                    rule.attrs.attribute = ctx.attrs.attribute
                end
                a = a .. stringifyAttrs(firstLineAttrs, rule.attrs)
                local a2 = stringifyAttrs(secondLineAttrs, rule.attrs)
                if #a2 ~= 0 then
                    a = a .. '\n' .. a2
                end
            end
            print(fmt('    "%s" [label="%s%s"];', escapedName, rule.tag, _escape:match(a)))

            if xmlBool(rule.attrs.lookAhead) then
                print(fmt('    "%s" [style=dashed];', escapedName))
            end


            if ruleContext == '#stay' then
                print(fmt('    "%s" -> "%s" [color=dodgerblue3];',
                          escapedName, escapedCtxName))
            elseif ruleContext then
                local escapedRuleContext = _escape:match(ruleContext)
                if ruleContext:byte(1) == sharp then
                    local escapedBindCtxName = _jumpctx:match(escapedRuleContext)
                    print(fmt('    "%s" -> "%s!!%s" [color=%s];\n    "%s!!%s" [label="%s"];',
                              escapedName,
                              escapedCtxName, escapedRuleContext, color,
                              escapedCtxName, escapedRuleContext, labelize(escapedRuleContext)))
                    if escapedBindCtxName then
                        kDot[escapedCtxName .. '!!' .. escapedRuleContext .. '!!' .. escapedBindCtxName] = {
                            fmt('  "%s!!%s" -> "%s" [color=%s];\n  "%s!!%s" [color=red];',
                                escapedCtxName, escapedRuleContext,
                                escapedBindCtxName, color,
                                escapedCtxName, escapedRuleContext),
                            irule,
                        }
                    end
                else
                    kDot[irule] = {
                        fmt('  "%s" -> "%s" [color=%s];',
                            escapedName, escapedRuleContext, color),
                        irule,
                    }
                end
            end
        end

        local fallthroughCtx = ctx.attrs.fallthroughContext
        local escapedFallthroughCtx = fallthroughCtx and _escape:match(fallthroughCtx)
        printContextAttr(escapedName, escapedCtxName, escapedFallthroughCtx, 'dashed', color)

        local endCtx = ctx.attrs.lineEndContext
        local escapedEndCtx = endCtx and _escape:match(endCtx)
        printContextAttr(escapedCtxName, escapedCtxName, escapedEndCtx, 'dotted', 'blue')

        print('  }')

        if fallthroughCtx then
            printLastTransition(escapedName, escapedCtxName, escapedFallthroughCtx, color)
        end

        if endCtx then
            printLastTransition(escapedName, escapedCtxName, escapedEndCtx, color)
        end

        for _,t in pairs(kIndexedTableToTable(kDot)) do
            print(t[2])
        end
    end
end
print('}')

File: utils/generate-dot-file.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: 2023 Jonathan Poelen <jonathan.poelen@gmail.com>
# SPDX-License-Identifier: MIT

from argparse import ArgumentParser
from pathlib import Path
from typing import Iterable, Mapping, Iterable
from textwrap import wrap
from xml.parsers import expat
from abc import ABC, abstractmethod

import re


parser = ArgumentParser(
    prog='generate-dot-file.py',
    description=f'''Dot file generator for xml syntax

Example:
    generate-dot-file.py data/syntax/lua.xml | dot -T svg -o image.svg''')

parser.add_argument('-c', '--context-only', action='store_true',
                    help='Generates contexts without rules')

parser.add_argument('-n', '--minimize-node', action='store_true',
                    help='Removes intermediate nodes that appear in the presence of #pop or multiple contexts (only with --context-only)')

parser.add_argument('-r', '--resolve-entities', action='store_true',
                    help='Evaluate xml entities')

parser.add_argument('-i', '--include', action='append', default=[],
                    help='Include only contexts that respect a pattern')

parser.add_argument('-e', '--exclude', action='append', default=[],
                    help='Exclude contexts that respect a pattern')

parser.add_argument('syntax', help='XML Syntax Definition Files')


args = parser.parse_args()

excludes = [re.compile(patt) for patt in args.exclude]
includes = [re.compile(patt) for patt in args.include]
context_only = args.context_only
minimize_node = args.minimize_node
resolve_entities = args.resolve_entities or context_only

global_entities = {
    '&#9;': '\\t',
    '&#37;': '%',
    '&#38;': '&',
    '&amp;': '&',
    '&#39;': "'",
    '&aquot;': "'",
    '&#34;': '"',
    '&quot;': '"',
    '&#60;': '<',
    '&lt;': '<',
    '&#62;': '>',
    '&gt;': '>',
}
entities_finder = re.compile('|'.join(global_entities))

Outside = 0
Context = 1
Rule = 2

class XMLParserBase(ABC):
    depth = Outside
    matched = False
    ictx = 0
    ctx_name = ''
    ctx_attrs: dict[str, str]
    escaped_ctx_name = ''
    ctx_color = ''
    irule = 0

    resolve_entities = True
    reversed_entities: dict[str, str] = {}
    resolved_entity_searcher: re.Pattern

    output: list[str]
    minimize_node: bool

    def __init__(self, output: list[str]):
        self.output = output
        self.ctx_attrs = {}
        self.reversed_entities = {}
        super().__init__()

    @abstractmethod
    def start_ctx(self) -> None:
        ...

    @abstractmethod
    def end_ctx(self) -> None:
        ...

    @abstractmethod
    def rule_process(self, irule: int, tag: str, attrs: dict[str, str]) -> None:
        ...

    def start_element(self, tag: str, attrs: dict[str, str]):
        if self.depth == Context:
            self.depth = Rule
            if self.matched:
                self.irule += 1
                if not self.resolve_entities:
                    string = attrs.get('String')
                    if string:
                        attrs['String'] = self.unresolve_entities(string)
                self.rule_process(self.irule, tag, attrs)
        elif tag == 'context':
            name = attrs['name']
            self.depth = Context
            self.matched = (not match_patterns(name, excludes)
                            and (not includes or match_patterns(name, includes)))
            if self.matched:
                self.irule = 0
                self.ctx_name = name
                self.ctx_attrs = attrs
                self.escaped_ctx_name = escape(name)
                self.ctx_color = compute_color(name)
                self.start_ctx()

    def end_element(self, name: str):
        if self.depth == Context:
            if self.matched:
                self.end_ctx()
            self.ictx += 1
        self.depth -= 1

    def unresolve_entities(self, s: str) -> str:
        """
        expat module converts all entities. This function tries to do the
        opposite by replacing pieces of text with entities.

        The result may differ from the original text, but will be equivalent.
        """
        b = True
        def replace(m):
            nonlocal b
            b = True
            return self.reversed_entities[m[0]]
        while b:
            b = False
            s = self.resolved_entity_searcher.sub(replace, s)
        return s

    def entity_decl(self, name, is_parameter_entity, value, base, system_id, public_id, notation_name):
        value = entities_finder.sub(lambda m: global_entities[m[0]], value)
        self.reversed_entities[value] = f'&{name};'

    def end_doctype(self):
        patt = '|'.join(re.escape(value) for value in self.reversed_entities)
        self.resolved_entity_searcher = re.compile(patt)


color_map = (
    '"/rdgy4/3"',
    '"/set312/1"',
    '"lightgoldenrod2"',
    '"/set312/3"',
    '"/set312/4"',
    '"/set312/5"',
    '"/set312/6"',
    '"/set312/7"',
    '"/rdpu3/2"',
    '"/purd6/3"',
    '"/ylgn4/2"',
    '"/set26/6"',
)

picked_colors: dict[int, str] = {}

def compute_color(name: str) -> str:
    """
    returns a color which depends on the first 2 characters
    """
    k = ord(name[0])
    if len(name) > 1:
        k += ord(name[1]) * 1024
    color = color_map[len(picked_colors) % len(color_map)]
    return picked_colors.setdefault(k, color)


def match_patterns(name: str, patterns: list[re.Pattern]) -> bool:
    return any(patt.search(name) for patt in patterns)


_pops_re = re.compile(r'^(?:#pop)+!?')

def split_ctx(ctx: str) -> tuple[str,  # '#pop' sequence
                                 str,  # seprator between pop and ctx
                                 str,  # contexts without pop
                                 Iterable[str]  # next contexts
                                 ]:
    pop = ''

    ctx_without_pop: str = _pops_re.sub('', ctx)
    n = len(ctx) - len(ctx_without_pop)
    if n:
        if n > 5:
            pop = f'#pop({n//4})'
        else:
            pop = '#pop'

    if ctx_without_pop:
        # unique context with order preservation
        ctxs = {ctx: None for ctx in ctx_without_pop.split('!')}
        sep = pop and ctx_without_pop and "!" or ""
        return (pop, sep, ctx_without_pop, ctxs.keys())

    return (pop, '', ctx_without_pop, ())


def stringify_attrs(attr_names: Iterable[str], attrs: Mapping[str, str]) -> str:
    s = ''
    for name in attr_names:
        attr = attrs.get(name)
        if attr:
            part = '\n'.join(wrap(attr, 40))
            s += f'  {name}:{part}'
    return s


def escape(s: str) -> str:
    return s.replace('\\', '\\\\').replace('"', '\\"')


def xml_bool(s: str | None) -> bool:
    return s == '1' or s == 'true'


class XMLContextOnlyParser(XMLParserBase):
    # avoid multi arrow for ctx1 -> ctx2
    krule_contexts: dict[str, int]
    minimize_node: bool

    def __init__(self, output: list[str], minimize_node: bool):
        self.krule_contexts = {}
        self.minimize_node = minimize_node
        super().__init__(output)

    def start_ctx(self) -> None:
        self.krule_contexts.clear()

    def rule_process(self, irule: int, name: str, attrs: dict[str, str]) -> None:
        self.krule_contexts[attrs.get('context') or '#stay'] = irule

    def end_ctx(self) -> None:
        color = self.ctx_color
        ctx_name = self.escaped_ctx_name
        self.output.append(f'  "{ctx_name}" [style=filled,color={color}]\n')

        self.krule_contexts.setdefault(self.ctx_attrs.get('fallthroughContext') or '#stay', -1)
        self.krule_contexts.setdefault(self.ctx_attrs.get('lineEndContext') or '#stay', -2)
        self.krule_contexts.setdefault(self.ctx_attrs.get('lineEmptyContext') or '#stay', -3)

        self.krule_contexts.pop('#stay')

        dejavu: set[str] = set()

        def push(s: str):
            if s not in dejavu:
                dejavu.add(s)
                self.output.append(s)

        for rule_context, i in sorted(self.krule_contexts.items(), key=lambda t: t[1]):
            escaped_rule_context = escape(rule_context)
            pop, sep, ctx_without_pop, next_contexts = split_ctx(escaped_rule_context)

            if self.minimize_node:
                for ctx in next_contexts:
                    push(f'  "{ctx_name}" -> "{ctx}" [color={color}];\n')
                continue

            # rule
            if i >= 0:
                style = ''
            # fallthroughContext
            elif i == -1:
                style = ',style=dashed'
            # lineEndContext
            elif i == -2:
                style = ',style=dotted'
            # lineEmptyContext
            else:  # if i == -3:
                style = ',style=bold'

            # create intermediate node
            if pop or len(next_contexts) > 1:
                node = f'{ctx_name}!!{escaped_rule_context}'
                push(
                    f'  "{ctx_name}" -> "{node}" [color={color}{style}];\n'
                    f'  "{node}" [label="{pop}{sep}{ctx_without_pop}"];\n'
                )
                if node not in dejavu:
                    dejavu.add(node)
                    for ctx in next_contexts:
                        self.output.append(f'  "{node}" -> "{ctx}" [color={color}];\n')
            # direct link
            else:
                push(f'  "{ctx_name}" -> "{escaped_rule_context}" [color={color}{style}];\n')


class XMLParser(XMLParserBase):
    # Cattribute is the context attribute if no attribute is specified
    # note: char1 is tranformed into String
    FIRST_LINE_ATTRIBUTES = ('attribute', 'Cattribute', 'String', 'char')
    SECOND_LINE_ATTRIBUTES = ('beginRegion', 'endRegion', 'lookAhead', 'firstNonSpace', 'column', 'additionalDeliminator', 'weakDeliminator', 'insensitive')

    kdot: dict[str, bool]
    escaped_name = ''

    def __init__(self, output: list[str]):
        self.kdot = {}
        super().__init__(output)

    def start_ctx(self) -> None:
        self.escaped_name = self.escaped_ctx_name

        self.kdot.clear()
        self.output.append(
            f'  subgraph cluster{self.ictx} {{\n'
            f'    "{self.escaped_name}" [shape=box,style=filled,color={self.ctx_color}];\n'
        )

    def rule_process(self, irule: int, name: str, attrs: dict[str, str]) -> None:
        color = self.ctx_color
        escaped_ctx_name = self.escaped_ctx_name

        escaped_name = f'{escaped_ctx_name}!!{irule}!!{escape(name)}'

        # context or rule linked to the next rule
        self.output.append(
            f'    "{self.escaped_name}" -> "{escaped_name}"'
            f' [style=dashed,color={color}];\n'
        )

        self.escaped_name = escaped_name

        rule_context = attrs.get('context', '#stay')

        if name == 'IncludeRules':
            label = f'  {rule_context}'
        else:
            if 'attribute' not in attrs:
                attrs['Cattribute'] = self.ctx_attrs['attribute']
            if 'char1' in attrs:
                attrs['String'] = attrs.pop('char') + attrs.pop('char1')
            label = stringify_attrs(self.FIRST_LINE_ATTRIBUTES, attrs)
            label2 = stringify_attrs(self.SECOND_LINE_ATTRIBUTES, attrs)
            if label2:
                label = f'{label}\n{label2}'
        self.output.append(f'    "{escaped_name}" [label="{name}{escape(label)}"];\n')

        if xml_bool(attrs.get('lookAhead')):
            self.output.append(f'    "{escaped_name}" [style=dashed];\n')

        if rule_context and rule_context != '#stay':
            self.push_link(escaped_name, rule_context, color, style='')
        else:
            self.output.append(
                f'    "{escaped_name}" -> "{escaped_ctx_name}" [color=dodgerblue3];\n'
            )

    def push_link(self, escaped_name: str, ctx: str, color: str, style: str) -> None:
        if style:
            style = f',style={style}'

        escaped_ctx = escape(ctx)
        pop, sep, ctx_without_pop, next_contexts = split_ctx(escaped_ctx)

        if pop or len(next_contexts) > 1:
            node = f'{self.escaped_ctx_name}!!{escaped_ctx}'
            self.kdot[f'    "{escaped_name}" -> "{node}" [color={color}{style}];\n'] = True
            self.kdot[f'    "{node}" [label="{pop}{sep}{ctx_without_pop}",color=red];\n'] = True
            for ctx in next_contexts:
                self.kdot[f'  "{node}" -> "{ctx}" [color={color}];\n'] = False
        else:
            self.kdot[f'  "{escaped_name}" -> "{escaped_ctx}" [color={color}{style}];\n'] = False

    def end_ctx(self) -> None:
        for escaped_ctx, attr, style, color in (
            (self.escaped_name, 'fallthroughContext', 'dashed', self.ctx_color),
            (self.escaped_ctx_name, 'lineEndContext', 'dotted', 'blue'),
            (self.escaped_ctx_name, 'lineEmptyContext', 'bold', 'purple'),
        ):
            ctx = self.ctx_attrs.get(attr)

            if ctx and ctx != '#stay':
                self.push_link(escaped_ctx, ctx, color, style)
            else:
                self.output.append(
                    f'    "{escaped_ctx}" -> "{self.escaped_ctx_name}"'
                    f' [style={style},color={color}];\n'
                )

        self.output.extend(s for s, in_cluster in self.kdot.items() if in_cluster)
        self.output.append('  }\n')
        self.output.extend(s for s, in_cluster in self.kdot.items() if not in_cluster)


output = [
    'digraph G {\n',
    '  compound=true;ratio=auto\n'
]

if context_only:
    xml_parser = XMLContextOnlyParser(output, minimize_node)
else:
    xml_parser = XMLParser(output)

p = expat.ParserCreate()
p.StartElementHandler = xml_parser.start_element
p.EndElementHandler = xml_parser.end_element
if not resolve_entities:
    xml_parser.resolve_entities = False
    p.EntityDeclHandler = xml_parser.entity_decl
    p.EndDoctypeDeclHandler = xml_parser.end_doctype

# # remove BOM
# if content.startswith('\xef\xbb\xbf'):
#     content = content[3:]
p.Parse(Path(args.syntax).read_text())

output.append('}\n')

print(''.join(output))
